reconcile/deadmanssnitch.py

```diff
@@ -0,0 +1,159 @@
+
+
+class DeadMansSnitchIntegration(QontractReconcileIntegration[NoParams]):
+    """Integration to automate deadmanssnitch snitch api during cluster dressup."""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def __init__(self) -> None:
+        super().__init__(NoParams())
+        self.qontract_integration_version = make_semver(0, 1, 0)
+        self.settings = get_deadmanssnitch_settings()
+        self.vault_client = cast(_VaultClient, VaultClient())
+
+    def write_snitch_to_vault(self, cluster_name: str, snitch_url: str) -> None:
+        self.vault_client.write(
+            {
+                "path": self.settings.snitches_path,
+                "data": {f"deadmanssnitch-{cluster_name}-url": snitch_url},
+            },
+            decode_base64=False,
+        )
+
+    def add_vault_data(self, snitch: Snitch, snitch_secret_path: str) -> Snitch:
+        try:
+            full_secret_path = {
+                "path": snitch_secret_path,
+                "field": f"deadmanssnitch-{snitch.get_cluster_name()}-url",
+            }
+            snitch.vault_data = self.secret_reader.read(full_secret_path).strip()
+        except SecretNotFound:
+            snitch.vault_data = SECRET_NOT_FOUND
+        return snitch
+
+    def get_current_state(
+        self,
+        deadmanssnitch_api: DeadMansSnitchApi,
+        clusters: list[ClusterV1],
+        snitch_secret_path: str,
+    ) -> dict[str, Snitch]:
+        # current state includes for deadmanssnithch response and associated secret in vault
+        snitches = deadmanssnitch_api.get_snitches(tags=["app-sre"])
+        # create snitch_map only for  the desired clusters
+        snitches_with_cluster_mapping = {
+            snitch.get_cluster_name(): snitch for snitch in snitches
+        }
+        current_state = {
+            cluster.name: self.add_vault_data(snitch, snitch_secret_path)
+            for cluster in clusters
+            if (snitch := snitches_with_cluster_mapping.get(cluster.name))
+        }
+        return current_state
+
+    def create_snitch(
+        self, cluster: ClusterV1, deadmanssnitch_api: DeadMansSnitchApi
+    ) -> None:
+        alert_email = [self.settings.alert_email]
+        payload = {
+            "name": cluster.prometheus_url,
+            "alert_type": self.settings.alert_type,
+            "interval": self.settings.interval,
+            "tags": self.settings.tags,
+            "alert_email": alert_email,
+            "notes": self.settings.notes_link,
+        }
+        snitch = deadmanssnitch_api.create_snitch(payload=payload)
+        self.write_snitch_to_vault(
+            cluster_name=cluster.name, snitch_url=snitch.check_in_url
+        )
+
+    def reconcile(
+        self,
+        dry_run: bool,
+        deadmanssnitch_api: DeadMansSnitchApi,
+        cluster: ClusterV1,
+        snitch: Optional[Snitch] = None,
+    ) -> None:
+        match (cluster.enable_dead_mans_snitch, snitch):
```

Comment:
i personally find this switch case confusing. consider a simple if-else?

Reply:
Original idea was the same..Can be done but involves lot of condition, like the following 
`
if cluster.enable_dead_mans_snitch and snitch is None:
   //create_snitch
if not cluster.enable_dead_mans_snitch and snitch:
//delete_snitch
if cluster.enable_dead_mans_snitch and snitch:
     if snitch.check_in_url != snitch.vault_value:
           //update in vault
`

Reply:
much more readable compared to `case (True, None)`

Reply:
will update it accordingly

Reply:
the condition is too nested, avoid multiple indented conditional check via extracting methods to give them meaningful names, add comments to explain each create/delete/update is a bad smell. Should be easy to read like this:


```python
if should_create_snitch(cluster, snitch):
   # self.create_snitch(
if should_delete_snitch(cluster, snitch):
   # self.delete_snitch(
if should_update_in_vault(cluster, snitch):
   # self.update_in_vault(
```

Reply:
saw the new change `snitch.needs_vault_update()`, it is good since it reduced 1 software cyclonic complexity, but if we step back to see the bigger scope here, we can find we always need to check `cluster.enable_dead_mans_snitch` and `snitch` together to do next action, which indicate we are missing a domain here.

There should be a domain that contains knowledge about the condition of each action and how to evaluate them, there are also cases that you need additional property such as clusters' prometheus but it's not available in gql generated model, `cluster_to_prometheus_mapping` is not suitable in current method because other args are all limited to a single object already, here it should also narrow down to a single prometheus.

Suppose you create a new domain model `ClusterSnitch`, contains all fields (cluster, snitch, prometheus...) you need, this method would be natural like this:

```python
  def reconcile(
        self,
        dry_run: bool,
        deadmanssnitch_api: DeadMansSnitchApi,
        cluster_snitch: ClusterSnitch,
  ) -> None:
    if cluster_snitch.need_create():
      # create
    if cluster_snitch.need_delete():
      # delete
    if cluster_snitch.need_update():
      # update
```